src/Dependencies/Collections/Internal/SegmentedHashSetEqualityComparer`1.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// NOTE: This code is derived from an implementation originally in dotnet/runtime:
+// https://github.com/dotnet/runtime/blob/v5.0.7/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSetEqualityComparer.cs
+//
+// See the commentary in https://github.com/dotnet/roslyn/pull/50156 for notes on incorporating changes made to the
+// reference implementation.
+
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Collections.Internal
+{
+    /// <summary>Equality comparer for hashsets of hashsets</summary>
+    internal sealed class SegmentedHashSetEqualityComparer<T> : IEqualityComparer<SegmentedHashSet<T>?>
+    {
+        public bool Equals(SegmentedHashSet<T>? x, SegmentedHashSet<T>? y)
+        {
+            // If they're the exact same instance, they're equal.
+            if (ReferenceEquals(x, y))
+            {
+                return true;
+            }
+
+            // They're not both null, so if either is null, they're not equal.
+            if (x == null || y == null)
+            {
+                return false;
+            }
+
+            var defaultComparer = EqualityComparer<T>.Default;
+
+            // If both sets use the same comparer, they're equal if they're the same
+            // size and one is a "subset" of the other.
+            if (SegmentedHashSet<T>.EqualityComparersAreEqual(x, y))
+            {
+                return x.Count == y.Count && y.IsSubsetOfHashSetWithSameComparer(x);
+            }
+
+            // Otherwise, do an O(N^2) match.
+            foreach (var yi in y)
+            {
+                var found = false;
+                foreach (var xi in x)
+                {
+                    if (defaultComparer.Equals(yi, xi))
+                    {
+                        found = true;
+                        break;
+                    }
+                }
+
+                if (!found)
+                {
+                    return false;
+                }
+            }
+
+            return true;
```

Comment:
What if `x` has every element in `y` and one additional element? That will pass the test above and fall through to `return true` even though the `x` has more elements.

Reply:
I'm not sure what the expected behavior here. This is not documented on docs.microsoft.com for `HashSet<T>.CreateSetComparer()`, and is the same behavior seen here:
https://github.com/dotnet/runtime/blob/6ca8c9bc0c4a5fc1082c690b6768ab3be8761b11/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSetEqualityComparer.cs#L25-L51

Reply:
This just seems like a bug

Reply:
https://github.com/dotnet/runtime/issues/69218

Can we reference that bug here? If they decide to fix that in their source we can follow up with a fix here.